### DISCUSSION TITLE
Any Rest API

### DISCUSSION THREAD

**Discussion by governoze** [supabase/supabase#26633] [Contribute to Supabase]
Can we use wrappers to treat any rest api as a foreign table?

Thanks

**Comment by TheOtherBrian1** [comment]
Foreign Data Wrappers in general can be designed to treat REST APIs as foreign tables. For instance, this project's Stripe FDW interacts with the payment service's REST API. 

I think Christophe Pettus, the CEO of PostgreSQL consultancy firm PGX, articulated the capabilities of FDWs the best:

> [FDWs are] a method for getting data from other data sources into Postgresql. [It] could be a different database, it could be a flat file, it could be the the National Weather Service weather feed, it can be any data source you want to include in inside your Postgres database...
> - [Christophe Pettus, PGCON 2023](https://youtu.be/7wuDJxpU7Fo?t=32)

The adaptability of FDWs has allowed real implementations to query from a diverse set of data sources over the years. [The PostgreSQL wiki ](https://wiki.postgresql.org/wiki/Foreign_data_wrappers#Generic_Web_Wrappers)contains plenty of examples of Foreign Data Wrappers that rely on NoSQL databases, flat files, and of course, REST APIs. 

Hopefully, this has provided you with clarity.

Can I ask what your use case is, if any? I may be able to point you towards resources that could be useful.

**Reply by governoze** [reply]
Sure! I am building a unified api for market data, so depending from the data you want for example Brazilian fixed income or mutual funds, you will be querying from our own data which we store on supabase postgresql, if it is us equity, our partner polygon is our font, so we would connect to their API. Many thanks for the help

**Reply by TheOtherBrian1** [reply]
If you just need to do basic CRUD operations, I personally would not recommend FDWs because of their complexity. Even simple ones can take up hundreds of lines of code. For reference, [the Stripe FDW is around 1000 lines of Rust code.
](https://github.com/supabase/wrappers/blob/main/wrappers/src/fdw/stripe_fdw/stripe_fdw.rs)

The primary benefit of FDWs is that they allow you to query foreign data, so if you plan on doing sorts, filtering, joins, etc. they can be quite useful, but even if it feels a bit roundabout, utilizing serverless functions and temporary tables as an intermediary between PostgreSQL and other data sources can be a smoother development experience, while still allowing for some level of realtime querying. 

If you'd prefer a means to actually upload your data to PostgreSQL via ETL, you could try [aribyte](https://airbyte.com/connectors/polygon-stock-api), a reputable and open source company that has a self-deployable instance for connecting different services, including Postgres and polygon. It is very thoroughly documented. However, if you really do need realtime advanced SQL features, then FDWs may be the most viable option. 

I would like to point you towards [Steampipe](https://hub.steampipe.io/plugins). It's an open source project, maintained by the [turbot team](https://turbot.com/) and written in GO, that helps developers create FDWs. The one caveat is that it currently only supports PostgreSQL 14 and lower. Awkwardly, SupaBase uses PostgreSQL 15. That could be a deal breaker. Steampipe, though, is more mature with a list of [premade connectors](https://hub.steampipe.io/plugins) that show off its capabilities. Most notably, it has [robust documentation](https://steampipe.io/docs/develop/writing-plugins) that can make the coding process less intimidating. Unfortunately, if you need a FDW that works with PostgreSQL 15, the only options that I know of are Supabase Wrappers and writing an FDW in C code. Writing an FDW in C is impractical without an intimate understanding of PostgreSQL's source code, so that really only leaves Wrappers.

Wrapper's Rust documentation is thorough, but it does not provide context for what wrappers are or how they work. For someone who is unfamiliar with FDW development, Wrapper's code will look intimidating and confusing. If you really want to explore Supabase Wrappers further, then I recommend watching [Christophe Pettus PGCon talk on FDWs](https://www.youtube.com/watch?v=7wuDJxpU7Fo&ab_channel=PGCon). Having this context will provide you with a greater sense of direction and reassurance. You could also try reading [Postgres's official documentation](https://www.postgresql.org/docs/current/fdwhandler.html), but I find that it's better as reference material than as a guide to understand how to implement PostgreSQL's features.

Hopefully, you found this helpful.

**Reply by TheOtherBrian1** [reply]
Piggybacking on what @burmecia said, you may also want to look into [pgsql-http](https://github.com/pramsey/pgsql-http), a Supabase supported extension that can make http requests to outside sources, such as APIs. Inside an PL/pgSQL function, you could utilize pgsql-http to request data from Polygon. Then you could format the return JSON into a temporary table so it can be used for complex queries.

This may not be the best solution, but as a proof of concept, implementing something like this could take as little as an hour and be overall promising.

**Reply by governoze** [reply]
Many thanks for all the reference and ideas. Will research and test a few and comeback with the solution! Thanks

**Comment by burmecia** [comment]
Normally the foreign table structure is closely related with API json schema, that is, a column maps to a field in the API json response. You can also have another approach, that is to define an universal jsonb column to store the whole API response as json, and then define another view based on that foreign table to decompose fields from the jsonb column. By doing that, you can implement a generic foreign table to ingest any REST API.
